tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+        let h = fileapi::CreateFileW(
+            addr.as_ptr(),
+            self.desired_access,
+            0,
+            attrs as *mut _,
+            fileapi::OPEN_EXISTING,
+            winbase::FILE_FLAG_OVERLAPPED | winbase::SECURITY_IDENTIFICATION,
+            ptr::null_mut(),
+        );
+
+        if h == handleapi::INVALID_HANDLE_VALUE {
+            return Err(io::Error::last_os_error());
+        }
+
+        let io = mio_windows::NamedPipe::from_raw_handle(h);
+        let io = PollEvented::new(io)?;
+
+        Ok(NamedPipe { io })
+    }
+}
+
+/// The pipe mode of a [NamedPipe].
+///
+/// Set through [NamedPipeOptions::pipe_mode].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum PipeMode {
+    /// Data is written to the pipe as a stream of bytes. The pipe does not
+    /// distinguish bytes written during different write operations.
+    ///
+    /// Corresponds to [PIPE_TYPE_BYTE][crate::winapi::um::winbase::PIPE_TYPE_BYTE].
+    Byte,
+    /// Data is written to the pipe as a stream of messages. The pipe treats the
+    /// bytes written during each write operation as a message unit. Any reading
+    /// function on [NamedPipe] returns [ERROR_MORE_DATA] when a message is not
+    /// read completely.
+    ///
+    /// Corresponds to [PIPE_TYPE_MESSAGE][crate::winapi::um::winbase::PIPE_TYPE_MESSAGE].
+    ///
+    /// [ERROR_MORE_DATA]: crate::winapi::shared::winerror::ERROR_MORE_DATA
+    Message,
+}
+
+/// Indicates the end of a [NamedPipe].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[non_exhaustive]
+pub enum PipeEnd {
+    /// The [NamedPipe] refers to the client end of a named pipe instance.
+    ///
+    /// Corresponds to [PIPE_CLIENT_END][crate::winapi::um::winbase::PIPE_CLIENT_END].
+    Client,
+    /// The [NamedPipe] refers to the server end of a named pipe instance.
+    ///
+    /// Corresponds to [PIPE_SERVER_END][crate::winapi::um::winbase::PIPE_SERVER_END].
+    Server,
+}
+
+/// Information about a named pipe.
+///
+/// Constructed through [NamedPipe::info].
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct PipeInfo {
+    /// Indicates the mode of a named pipe.
+    pub mode: PipeMode,
+    /// Indicates the end of a named pipe.
+    pub end: PipeEnd,
+    /// The maximum number of instances that can be created for this pipe.
+    pub max_instances: u32,
+    /// The number of bytes to reserve for the output buffer.
+    pub out_buffer_size: u32,
+    /// The number of bytes to reserve for the input buffer.
+    pub in_buffer_size: u32,
+}
+
+/// Information about a pipe gained by peeking it.
+///
+/// See [NamedPipe::peek].
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct PipePeekInfo {
+    /// Indicates the total number of bytes available on the pipe.
+    pub total_bytes_available: usize,
+    /// Indicates the number of bytes left in the current message.
+    ///
+    /// This is undefined unless the pipe mode is [PipeMode::Message].
```

Comment:
According to [this page](https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-peeknamedpipe), it is set to zero for non-message pipes.
```suggestion
    /// If the pipe mode is not [PipeMode::Message], then this is zero.
```

Reply:
Thanks!